### PR TITLE
Add `SetNonNullable` type

### PR DESCRIPTION
**Pull request by Smtih**
Adds `SetNonNullable<Type, Keys>` which allows someone to remove nullish values at specific keys of an object. This is useful when defining type guards.

```ts
type Person {
	firstName?: string | null;
	lastName?: string | null;
}

function hasFirstName(person: Person) person is SetNonNullable<Person, "firstName"> {
	return person.firstName != null;
}
```

Continuation of https://github.com/sindresorhus/type-fest/pull/199
Addresses https://github.com/sindresorhus/type-fest/issues/228

If anything is still missing let me know, I've tried to fulfil the contributing guidelines.

### CODE REVIEW

File: source/set-non-nullable.d.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import {Except} from './except';
+import {Simplify} from './simplify';
+
+/**
+Create a type that removes nullish (optional, undefined and null) types from the given keys. The remaining keys are kept as is.
+This behaves in a similar manor to `SetRequired<T,K>` by removing optionals, however also applies the `NonNullable<T>` operator to the values at the specified keys.
```
Comment (BendingBender) [inline source/set-non-nullable.d.ts:6]:
```suggestion
Create a type that removes nullish (optional, `undefined` and `null`) types from the given keys. The remaining keys are kept as is.
This behaves in a similar manner to `SetRequired<T,K>` by removing optionals, however also applies the `NonNullable<T>` operator to the values at the specified keys.
```

File: source/set-non-nullable.d.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	a?: number;
+	b: string;
+	c?: boolean;
+	d: number | null;
+	e: string | undefined;
+}
+
+type SomeNonNullable = SetNonNullable<Foo, 'b' | 'c' | 'd' | 'e'>;
+// type SomeNonNullable = {
+// 	a?: number;
```
Comment (BendingBender) [inline source/set-non-nullable.d.ts:24]:
```suggestion
// 	a?: number; // Kept as is, was not mentioned in `'b' | 'c' | 'd' | 'e'` type argument.
```

**Review by BendingBender — CHANGES_REQUESTED** [review]
Other than the suggested changes this looks good to me. I have just one principled question: Wouldn't it make more sense to make this type have only one responsibility, just like `SetRequired`? I mean that this type should only remove the `null`/`undefined` from the object values but don't touch the optionality of the keys? This could make a difference especially in the context of the newly introduced [`--exactOptionalPropertyTypes` compiler option](https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#exact-optional-property-types). When designed like this, users could mix and match `SetNonNullable`/`SetRequired` as they see fit and we could even introduce a combined type that makes both transformations. @sindresorhus

**Comment by sindresorhus** [comment]
I agree with @BendingBender. The name is also misleading as it contains `NonNullable` in the name, but it does even more than `NonNullable`.

**Comment by sindresorhus** [comment]
@Smtih If you have time, it would be great if you could help review some of the other open pull requests. 🙏

**Comment by goodoldneon** [comment]
+1 to the name concerns. Maybe `SetNonNullish` instead?

File: source/set-non-nullable.d.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+// }
+
+function hasRequiredKeys(foo: Foo): foo is SomeNonNullable;
+```
+
+This is especially useful in cases where you want to write a type-guard that asserts one or more values of an object are not null.
+
+@example
+```
+type Person {
```
Comment (goodoldneon) [inline source/set-non-nullable.d.ts:38]: Missing `=`

File: source/set-non-nullable.d.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	firstName?: string | null;
+	lastName?: string | null;
+}
+
+function hasFirstName(person: Person) person is SetNonNullable<Person, "firstName"> {
```
Comment (goodoldneon) [inline source/set-non-nullable.d.ts:43]:
```suggestion
function hasFirstName(person: Person): person is SetNonNullable<Person, "firstName"> {
```

**Comment by sindresorhus** [comment]
> Maybe SetNonNullish instead?

I don't think the distinction between "nullable" and "nullish" are clear enough.

**Comment by goodoldneon** [comment]
> I don't think the distinction between "nullable" and "nullish" are clear enough.

“Nullish” is the official term for “either `null` or `undefined`“. For example:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator

**Comment by sindresorhus** [comment]
@goodoldneon Yes, but TS uses "nullable" for null/undefined in `NonNullable`. So it just makes it slightly different without actually making it any clearer that it also deals with optionality.

**Comment by goodoldneon** [comment]
@sindresorhus good point. It seems like the JavaScript community considers a "nullable" to be a something that can accept either `null` or `undefined`. It's unfortunate that `null` and `undefined` are conflated like that, but it is what it is.

If the goal of the proposed type is to apply the standard library `NonNullable` type over specific keys, then I think `SetNonNullable` is the most intuitive option

**Comment by sindresorhus** [comment]
@Smtih Still interested in finishing this?

**Pull request closed without merging**